### ISSUE TITLE
New problem with circular references

### ISSUE THREAD

**Issue by DoctorGM** [jonasthewolf/gsn2x#52]
Hi Jonas,
many thanks for the 2.1 update. My test file with the circular reference caused an error message. However, the error message was not related to the circular reference - it complained about not having a undeveloped node, which is not what I would have expected.

So I added an undeveloped node and the error went away, despite there still being a circular reference. My new test file is...

G1:
  text: This is a Goal
  supportedBy: [S1]

S1:
  text: This is a Strategy.
  supportedBy: [G2.1]
  
G2.1:
    text: subgoal
    # circular reference!!
    supportedBy: [G1]

G2.2:
    text: unreferenced
    undeveloped: true
   
Sorry... :-(

Very best regards...
gary

**Comment by jonasthewolf** [comment]
Hi @DoctorGM, The error message should have been about "C01: no unreferenced element". Because the program cannot know where the circle actually starts. Thus, it cannot actually say where the circle is.

**Comment by DoctorGM** [comment]
Hi Jonas,
you are correct. I did not remember the message correctly. However, the circular reference still remains in my example input YAML. Also, the error message could be improved a little...

Very best regards...
gary

**Comment by jonasthewolf** [comment]
I saw what you did: you created two completely unrelated graphs, one without edges and one with a circular reference...
Let's see if I can detect that...
At the first glance, it does not seem too easy.

**Comment by DoctorGM** [comment]
Hi Jonas,
I think it would be useful if the user was given some clue about where to look for the error. Assuming that it's a circular reference then printing out the entire cycle(s) would be the most useful but also, probably, the most painful.
I guess that you start scanning at a particular node in order to detect the cycle. So could you print out that there's a cycle and state which node it starts at?

Sorry to be a pain...
gary

**Comment by DoctorGM** [comment]
Hi Jonas,
Regarding
"I saw what you did: you created two completely unrelated graphs"
Yes... I cheated a little... :-) I would guess that the easiest solution here is to state (in the manual) that a file can contain only one graph and then have a test - that each node belongs to the same graph. Make it a very early test and if that test fails just exit without doing any further tests that rely on a single graph structure. Only a suggestion, though. You know the code much better than I do.

Have a good day...
gary

PS I urgently need to work on some safety documents. So my responses will slow down for the rest of the week.
